managed-pointer/cpp-20/D0567R1.md

```diff
@@ -229,94 +391,105 @@ auto gpuExecutor = gpuContext.executor();
 
-/* Construct a managed_ptr ptraA allocated on the host CPU */
-std::experimental::execution::managed_ptr<float> ptrA(1024);
+For example:
 
-/* Construct a managed_ptr ptrB allocated on the GPU execution context */
-std::experimental::execution::managed_ptr<float> ptrB(1024, gpuAllocator);
+```cpp
+using namespace std::experimental::concurrency_v2::execution;
 
-/* Populate ptrA */
-populate(ptrA);
+struct my_data { /* ... */ };
 
-/* Perform the compute operation with implicit synchronisation operations to the GPU execution context */
-std::experimental::execution::async(gpuExecutor, [=](auto _ptrA, auto _ptrB) { /* ... */ }, ptrA, ptrB).wait();
+vendor_x::input_stream_execution_context inputStreamContext;
+vendor_x::gpu_execution_context gpuContext;
+vendor_x::input_stream_to_gpu_channel inputStreamToGPUChannel;
 
-/* Perform a synchronisation operation to the host CPU */
-ptrB.get().wait();
+auto inputStreamExec = inputStreamContext.executor();
 
-/* Print the result */
-print(ptrB);
+auto gpuExec = gpuContext.executor();
+
+managed_ptr<my_data> ptr(inputStreamContext.global_memory_allocator());
+
+for (;;) {
+  auto fut =
+    execute(inputStreamExec, [=](){ read_input(ptr); })
+      .then_sync(inputStreamToGPUChannel, inputStreamExec, gpuExec)
+        .then_execute(gpuExec, [=](){ process(ptr); })
+  post_process(fut.get());
+}
 ```
+*Figure 3: Synchronisation Channels Example*
+
+> This is not the most efficient example, it could be improved by double buffering the sycnhornisation and computation, though it demonstrates how channels can be used to avoid sycnhronising via a host.
+
+## 4. Use Cases
 
-*Figure 4: Example of implicit interface*
+// TODO
 
-## Future Work
+## 5. Future Work
 
 There are many other considerations to make when looking at a model for data movement for heterogeneous and distributed systems, however, this paper aims to establish a foundation which can be extended to include other paradigms in the future.
 
-### More Complex Data Movement Policies
+### 5.1. Shared Accessibility
 
-We may wish to introduce more control over the way in which data is moved between devices at a high level and the consistency between execution contexts. This could be done by having static type traits associated with executor classes which describe the data movement and consistency properties of the execution context associated with the executor.
+Systems which support a unified address space (i.e. sharing a single memory region across different *execution contexts*) could allow for a `managed_ptr` instance to share accessibility between multiple *execution contexts*. This would allow a `managed_ptr` to be concurrently accessible by multiple *execution contexts* via the use of atomics providing those *execution contexts* share a memory region.
```

Comment:
This is the question that will need to be answered. On the one hand you could say no because if you have a system with a unified address space then you can likely just pass a raw pointer directly. However if you wish to inter-operate with executors in systems which don't, or if you want to write generic code to support a range of different systems which may nor may not have a unified address space, or if you want manage the ownership of pointers between executors then you may still want to use a `managed_ptr`.